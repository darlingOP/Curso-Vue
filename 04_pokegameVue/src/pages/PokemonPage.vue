<template>
    <h1 v-if="!pokemon">Espere por favor...</h1>

    <div v-else>
        <h1>¿Quíen es este pokémon</h1>

        <poke-img 
        :pokeId="pokemon.id" 
        :showPokemon="showPokemon"/>

        <poke-opts 
        :options="pokemonArr"
        @selection="checkAnswer"/>

        <template v-if="showAnswer">
            <h2 class="fade-in">{{ message }}</h2>
            <button @click="newGame">
                Nuevo Juego
            </button>
        </template>
        
    </div>
   
</template>
<script>
import PokeImg from '@/components/PokemonImage.vue'
import PokeOpts from '@/components/PokemonOptions.vue'

import getPokemonOptions from '@/helpers/getPokemonOptions'

export default {
    components:{
        PokeImg,
        PokeOpts
    },
    data(){
        return{
            pokemonArr:[],
            pokemon:null,
            showPokemon:false,
            showAnswer:false,
            message:''
        }
    },
    methods: {
        async mixPokemonArray(){

            this.pokemonArr = await getPokemonOptions()

            const rndInt = Math.floor(Math.random() * 4)
            this.pokemon = this.pokemonArr[rndInt]
        },

        checkAnswer(selectedId){
            this.showPokemon = true
            this.showAnswer = true
            if(selectedId === this.pokemon.id){
                this.message = `Correcto, era ${this.pokemon.name}`
            }else{
                this.message = `Oops!... era ${this.pokemon.name}`
            }
        },
        newGame(){
            this.showPokemon = false
            this.showAnswer = false
            this.pokemonArr = []
            this.pokemon = null
            this.mixPokemonArray()
        }
    },
    mounted(){
        this.mixPokemonArray()
    }
}
</script>
<style>
    
</style>