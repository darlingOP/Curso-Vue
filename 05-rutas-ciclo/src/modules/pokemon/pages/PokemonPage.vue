<template>
    <h1>
        Pokemon page <span v-if="pokemon">#{{ id }}</span> 
    </h1>
    <img v-if="pokemon" :src="pokemon.sprites.front_default" :alt="pokemon.name">
</template>

<script>
    export default {
        props:{
            id:{
                type:Number,
                required:true
            }
        },
        data(){
            return{
               pokemon:null
            }
        },       
        created(){
            //se recomienda que las peticiones sean llamadas en created
            this.getPokemon()
        },

        methods:{

            async getPokemon(){

                try {
                    const pokemon = await fetch(`https://pokeapi.co/api/v2/pokemon/${this.id}`)
                    .then(r => r.json()) 
                    console.log(pokemon);
                    this.pokemon = pokemon  
                } catch (error) {
                    this.$router.push('/') //metodo brindado por router
                    console.log('no hay nada que hacer aqui'); 
                }
               
            }
        },

        watch:{
            id(){
                this.getPokemon()
            }
        }
    }
</script>

<style scoped>
    
</style>