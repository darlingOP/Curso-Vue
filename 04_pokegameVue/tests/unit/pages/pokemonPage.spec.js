import { shallowMount, mount } from "@vue/test-utils";
import PokemonPage from '@/pages/PokemonPage'
import { mockPokemons } from "../mocks/pokemons.mock";

describe('PokemonPage Component',() =>{
    let wrapper

    beforeEach(() =>{
        wrapper = shallowMount(PokemonPage)
    })

    test('debe llamar al pokemonArray al montar ', () => {
        const mixPokemonArraySpy = jest.spyOn(PokemonPage.methods,'mixPokemonArray')
        const wrapper = shallowMount(PokemonPage)
        expect(mixPokemonArraySpy).toHaveBeenCalled()
    });

    test('debe hacer match con el snapshot cuando se monta el componente', () => {

        const wrapper = shallowMount(PokemonPage,{
            data(){
                return{
                    pokemonArr:[mockPokemons],
                    pokemon:mockPokemons[0],
                    showPokemon:false,
                    showAnswer:false,
                    message:''
                }
            }
        })

        expect( wrapper.html()).toMatchSnapshot()

    });

    test('debe de mostrar los componentes pokemonImage y pokemonOptions', () => {
        
        //pokemonImage existe
        //pokemonOptions existe

        //pokemonImage attribute pokeId === 5
        //pokemonOptions attribute options existe
        const wrapper = shallowMount(PokemonPage,{
            data(){
                return{
                    pokemonArr:[mockPokemons],
                    pokemon:mockPokemons[0],
                    showPokemon:false,
                    showAnswer:false,
                    message:''
                }
            }
        })

        //se busca el componente mediante el nombre que tiene en el snapshot creado con el shallowMount
        //ya que es una simulacion de los componentes reales
        const image = wrapper.find('poke-img-stub')
        const options = wrapper.find('poke-opts-stub')
        
        expect(image.exists()).toBeTruthy()
        expect(options.exists()).toBeTruthy()

        expect(image.attributes('pokeid')).toBe('5')
        expect(options.attributes('options')).toBeTruthy()

    });

    test('pruebas con checkanswer', async () => {
        const wrapper = shallowMount(PokemonPage,{
            data(){
                return{
                    pokemonArr:[mockPokemons],
                    pokemon:mockPokemons[1],
                    showPokemon:false,
                    showAnswer:false,
                    message:''
                }
            }
        })

        //se debe esperar a que vue realice la renderizacion
         await wrapper.vm.checkAnswer(10)

        expect(wrapper.find('h2').exists).toBeTruthy()
        expect(wrapper.vm.showPokemon).toBe(true)
        expect(wrapper.find('h2').text()).toBe(`Correcto, era ${mockPokemons[1].name}`)

        await wrapper.vm.checkAnswer(5)
        expect(wrapper.vm.message).toBe(`Oops!... era ${mockPokemons[1].name}`)
    });
})