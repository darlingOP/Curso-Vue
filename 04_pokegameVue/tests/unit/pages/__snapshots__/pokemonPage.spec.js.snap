// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PokemonPage Component debe hacer match con el snapshot cuando se monta el componente 1`] = `
<div>
  <h1>¿Quíen es este pokémon</h1>
  <poke-img-stub pokeid="5" showpokemon="false"></poke-img-stub>
  <poke-opts-stub options="[object Object],[object Object],[object Object],[object Object]"></poke-opts-stub>
  <!--v-if-->
</div>
`;
